public/pages/utils/anomalyResultUtils.ts

```diff
@@ -483,3 +523,288 @@ export const parsePureAnomalies = (
   }
   return anomalies;
 };
+
+export type FeatureDataPoint = {
+  isMissing: boolean;
+  plotTime: number;
+  startTime: number;
+  endTime: number;
+};
+
+export const FEATURE_DATA_CHECK_WINDOW_OFFSET = 2;
+
+export const getFeatureDataPoints = (
+  featureData: FeatureAggregationData[],
+  interval: number,
+  dateRange: DateRange | undefined
+): FeatureDataPoint[] => {
+  const featureDataPoints = [] as FeatureDataPoint[];
+  if (!dateRange) {
+    return featureDataPoints;
+  }
+
+  const existingTimes = isEmpty(featureData)
+    ? []
+    : featureData
+        .map(feature => getFloorPlotTime(feature.startTime))
+        .filter(featureTime => featureTime != undefined);
+  for (
```

Comment:
The time complexity of this block is O(m*n). m is intervals, n is existingTimes count. Have you tested the performance? Suggest to use binary search in function `findTimeExistsInWindow` as existingTimes is ordered.

Reply:
currently with 7 days data performance seems ok. will change to use binary search or other improvement for `findTimeExistsInWindow`